Fix vkb::screenshot()

gary-sweet opened the pull request:
## Description

Addresses the incorrect use of VK_IMAGE_USAGE_SAMPLED_BIT as described in issue #259.

Changes vkb::screenshot() to use VkCopyImageToBuffer  rather than VkCopyImage.

This simplifies the code, and I believe is a better practice that trying to read image data directly.
The previous code also didn't work for me. It either produced empty images, or images that started with a horizontal offset for me.

Tested only on Linux devices.

Fixes #259

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [x] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [x] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] My changes do not add any regressions
- [x] I have tested every sample to ensure everything runs correctly
- [x] This PR describes the scope and expected impact of the changes I am making

gary-sweet committed:
Fix vkb::screenshot()

See https://github.com/KhronosGroup/Vulkan-Samples/issues/259

As well as addressing the incorrect use of VK_IMAGE_USAGE_SAMPLED_BIT
as described in the issue, this change modifies the screenshot()
function to use VkCopyImageToBuffer rather than VkCopyImage as
it was using before. This results in both simpler and working
code (for me at least). The old version either produced empty
images, or images that started with a horizontal offset for me.

gpx1000 commented on framework/common/utils.cpp:
```diff
@@ -58,37 +58,31 @@ void screenshot(RenderContext &render_context, const std::string &filename)
 	auto &frame          = render_context.get_last_rendered_frame();
```
the local variable format define at line 55 is no longer used

We also no longer use the following includes:
#include "core/pipeline_layout.h"
#include "core/shader_module.h"
#include "scene_graph/components/image.h"
#include "scene_graph/components/mesh.h"
#include "scene_graph/components/pbr_material.h"
#include "scene_graph/components/sampler.h"
#include "scene_graph/components/texture.h"
#include "scene_graph/components/transform.h"

gary-sweet replied:
Done

gpx1000 commented on framework/common/utils.cpp:
```diff
@@ -109,26 +103,26 @@ void screenshot(RenderContext &render_context, const std::string &filename)
 	swizzle          = std::find(bgr_formats.begin(), bgr_formats.end(), src_image_view.get_format()) != bgr_formats.end();
```
it makes sense to just define swizzle here as line 99 definition with setting to false is never read and proves unused until here.

gary-sweet replied:
Agreed. This was the original code, but does look a bit odd. I'll fix it up.

gary-sweet replied:
Done

gpx1000 commented on framework/core/command_buffer.h:
```diff
@@ -1,4 +1,4 @@
+/* Copyright (c) 2019-2021, Arm Limited and Contributors
  *
  * SPDX-License-Identifier: Apache-2.0
  *
```
line 81: move constructors should be marked noexcept

line 161: End of the line has an extra ';'.  The macro LOGE handles the ';'

gary-sweet replied:
Neither of those lines have been modified by this PR, and similar issues likely exist in many other places, so I consider them beyond the scope of this change.

gpx1000 commented on framework/core/command_buffer.cpp:
```diff
@@ -455,6 +455,12 @@ void CommandBuffer::copy_buffer_to_image(const core::Buffer &buffer, const core:
 	                       to_u32(regions.size()), regions.data());
 }
 
+void CommandBuffer::copy_image_to_buffer(const core::Image &image, VkImageLayout image_layout, const core::Buffer &buffer, const std::vector<VkBufferImageCopy> &regions)
```
this method can be made const.

gary-sweet replied:
It could, but doing so would make it inconsistent with all the other similar methods around it. In general, I prefer to fit in with the existing code (unless it's blatantly wrong of course). If we were to make this const, then we'd also want to make a whole lot more of these const, and that's well beyond the scope of this fix.

gpx1000 reviewed (dismissed):
Both the old version of using a screenshot and this method work in Linux when tested locally.  I think we should change the key binding from printscr to somehting else as that key is bound by the system in linux to capture the whole screen.

The comments I found are minor and shouldn't hold up the acceptance of this PR.  If there's time it'd be great to address them otherwise, this LGTM;

gary-sweet committed:
Addressed some review comments

Added F12 as an alternative screendump trigger.
Removed some unused includes.
A couple of code tidies.

gary-sweet committed:
Updated copyright year

SaschaWillems commented:
LGTM and works as expected on an NVIDIA RTX 2060.

One thing I noticed, which is not related to this PR at all, is that the screenshot feature only seems to work in samples using the ARM framework, but not in the samples using the API sample base class as the input handler is never called in these. That may be worth a separate issue.

ZandroFargnoli reviewed (approved):
LGTM! Tested on Windows 👍

marty-johnson59 commented:
Assuming Steven is still OK with this (from comments, it looks like he is), this looks like it's ready to merge.
